File: python/reverse_linked_list_ii.py
"""
Definition of ListNode

class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The head of linked list
    @param m: start position
    @param n: end position
    """
    def reverseBetween(self, head, m, n):
        if not head:
            return None
        # as the head might be reversed, use a dummy node to keep tracking
        dummy = ListNode(-1)
        dummy.next = head
        
        # before reversing, iterate over until the previous of the target
        prev = dummy
        for i in xrange(1, m):
            if not prev:
                return None
            prev = prev.next
        
        # now we can start reversing
        curr = prev.next
        next = curr.next
        for i in xrange(m, n):
            curr.next = next.next
            next.next = prev.next
            prev.next = next
            next = curr.next
            
        return dummy.next

File: java/ReverseLinkedListII.java
public class ReverseLinkedListII {
    /**
     * @param ListNode head is the head of the linked list
     * @oaram m and n
     * @return: The head of the reversed ListNode
     */
    public ListNode reverseBetween(ListNode head, int m , int n) {
        if (head == null || head.next == null || m >= n) {
            return head;
        }

        ListNode dummy = new ListNode(0);
        ListNode prev = dummy;
        dummy.next = head;

        for (int i = 1; i < m; i++) {
            prev = prev.next;
        }
        ListNode start = prev.next;

        ListNode end = head;
        for (int i = 1; i < n; i++) {
            end = end.next;
        }

        prev.next = reverse(start, end, n - m + 1);

        return dummy.next;
    }

    public ListNode reverse(ListNode start, ListNode end, int len) {
        ListNode prev = end.next;
        ListNode curr = start;
        while (len != 0) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
            len--;
        }
        return prev;
    }
}
